holoviews/core/data/grid.py

```diff
@@ -189,11 +189,11 @@ def groupby(cls, dataset, dim_names, container_type, group_type, **kwargs):
         group_kwargs.update(kwargs)
 
         # Find all the keys along supplied dimensions
-        keys = [dataset.data[d.name] for d in dimensions]
+        keys = [dataset.dimension_values(d.name, False) for d in dimensions]
 
         # Iterate over the unique entries applying selection masks
         grouped_data = []
-        for unique_key in zip(*util.cartesian_product(keys)):
+        for unique_key in zip(*(arr.flat for arr in util.cartesian_product(keys))):
```

Comment:
The definition of ``cartesian_product(arrays):`` is simply  ``np.broadcast_arrays(*np.ix_(*arrays))``. I'm wondering if it might be useful behavior to always output an iterator over flattened/flat arrays...